### PR TITLE
More bug fixes

### PR DESCRIPTION
**Pull request by brubraz**
### Features
- **Qi'ra and Viper Droid**: Added a new pop-up to look at the opponent's hand.  

### Fixes
- Resolved index issues in the `GetAllyWhenDestroyTheirsEffects`. 
- ~~**Ezra**: Resolved issues when a unit completes an attack.~~ (Fixed by @mariotee in #448 )
- **Snoke**:  
  - Resolved index issues when defeating enemies without HP.  
  - Defeat tokens if they have no remaining HP.  
- **Resolute**: Resolved index issues.  
- **Foresight**: Updated to retrieve the top deck asynchronously.

### CODE REVIEW

File: CardDictionary.php [hunk 1]
```diff
@@ -1505,8 +1505,8 @@ function IsToken($cardID)
   switch($cardID) {
     case "8752877738": return true;
     case "2007868442": return true;
-    case "3463348370": return true;
-    case "3941784506": return true;
+    case "3463348370": return true;//Battle Droid
```
Comment (mariotee) [inline CardDictionary.php:1508]: i love this refactor

File: Libraries/NetworkingLibraries.php [hunk 2]
```diff
@@ -881,6 +881,7 @@ function ResolveSingleTarget($mainPlayer, $defPlayer, $target, $attackerPrefix,
   global $combatChain, $combatChainState, $mainPlayer, $defPlayer, $CCS_CombatDamageReplaced;
   global $CCS_DamageDealt;
 
+  $attackerSurvived = 1;
```
Comment (mariotee) [inline Libraries/NetworkingLibraries.php:884]: you can remove this and check `!$attackerDestroyed` below (~line 939)

File: AllyAbilities.php [hunk 3]
```diff
@@ -40,18 +40,27 @@ function PlayAlly($cardID, $player, $subCards = "-", $from = "-", $owner = null,
...
     $allies = &GetAllies($player);
-    $prefix = $currentPlayer == $player ? "MYALLY-" : "THEIRALLY-";
     for ($i = 0; $i < count($allies); $i += AllyPieces()) {
-      $ally = new Ally($prefix . $i, $player);
-      $ally->DefeatIfNoRemainingHP();
+      $ally = new Ally("MYALLY-" . $i, $player);
+      $defeated = $ally->DefeatIfNoRemainingHP();
+
+      if ($defeated) {
+        $i -= AllyPieces(); // Decrement to account for the removed ally
```
Comment (mariotee) [inline AllyAbilities.php:62]: nice

File: Libraries/NetworkingLibraries.php (hunk 2 above)
Comment (mariotee) [reply Libraries/NetworkingLibraries.php:884]: oh also can initialize `$attackerDestroyed = 0` here. i had a similar fix for the Ezra bug in #448

Comment (brubraz) [reply Libraries/NetworkingLibraries.php:884]: @mariotee Did you fix this in #448, right? It's related to Ezra's bug. So, I'll remove my changes.

**Pull request merged**